src/http/h1/decode.rs

```diff
@@ -87,183 +93,248 @@ impl Decoder {
-
-                }
-            },
-            // If we weren't in the extension yet, the ";" signals its start
-            b';' if !in_ext => {
-                in_ext = true;
-                in_chunk_size = false;
-            },
-            // "Linear white space" is ignored between the chunk size and the
-            // extension separator token (";") due to the "implied *LWS rule".
-            b'\t' | b' ' if !in_ext & !in_chunk_size => {},
-            // LWS can follow the chunk size, but no more digits can come
-            b'\t' | b' ' if in_chunk_size => in_chunk_size = false,
-            // We allow any arbitrary octet once we are in the extension, since
-            // they all get ignored anyway. According to the HTTP spec, valid
-            // extensions would have a more strict syntax:
-            //     (token ["=" (token | quoted-string)])
-            // but we gain nothing by rejecting an otherwise valid chunk size.
-            _ext if in_ext => {
-                //TODO: chunk extension byte;
-            },
-            // Finally, if we aren't in the extension and we're reading any
-            // other octet, the chunk size line is invalid!
+            b @ b'0'...b'9' => {
+                *size *= radix;
+                *size += (b - b'0') as u64;
+            }
+            b @ b'a'...b'f' => {
+                *size *= radix;
+                *size += (b + 10 - b'a') as u64;
+            }
+            b @ b'A'...b'F' => {
+                *size *= radix;
+                *size += (b + 10 - b'A') as u64;
+            }
+            b'\t' | b' ' => return Ok(ChunkedState::SizeLws),
+            b';' => return Ok(ChunkedState::Extension),
+            b'\r' => return Ok(ChunkedState::SizeLf),
             _ => {
                 return Err(io::Error::new(io::ErrorKind::InvalidInput,
-                                         "Invalid chunk size line"));
+                                          "Invalid chunk size line: Invalid Size"));
             }
         }
+        Ok(ChunkedState::Size)
     }
-    trace!("chunk size={:?}", size);
-    Ok(size)
-}
+    fn read_size_lws<R: Read>(rdr: &mut R) -> io::Result<ChunkedState> {
+        trace!("read_size_lws");
+        match byte!(rdr) {
+            // LWS can follow the chunk size, but no more digits can come
+            b'\t' | b' ' => Ok(ChunkedState::SizeLws),
+            b';' => Ok(ChunkedState::Extension),
+            b'\r' => return Ok(ChunkedState::SizeLf),
+            _ => {
+                Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                   "Invalid chunk size linear white space"))
+            }
+        }
+    }
+    fn read_extension<R: Read>(rdr: &mut R) -> io::Result<ChunkedState> {
+        trace!("read_extension");
+        match byte!(rdr) {
+            b'\r' => return Ok(ChunkedState::SizeLf),
+            _ => return Ok(ChunkedState::Extension), // no supported extensions
```

Comment:
Pretty crude for now. If we receive a `;` in the size line, we just keep reading until the next `\r`.